Implement detector covers

mikerooni opened the pull request:
- [x] Activity
- [x] Energy
- [x] Fluid
- [x] Item
- [x] Maintenance

mikerooni committed:
feat: add redstone output to cover behavior

mikerooni committed:
style: Reformat cover & cover item definitions

mikerooni committed:
fix: covers now output redstone to the correct side

mikerooni committed:
feat: add activity detector covers

mikerooni committed:
refactor: move workable out of DetectorCover base class

mikerooni committed:
feat: add IMaintenanceMachine capability

mikerooni committed:
feat: maintenance detector cover

mikerooni committed:
fix: don't use circuit in maintenance detector cover recipe

mikerooni committed:
Merge branch '1.19.2' into detector-covers

mikerooni committed:
fix: correctly handle UI covers in DetectorCover base class

mikerooni committed:
feat: energy detector covers

mikerooni committed:
feat: item detector covers

mikerooni committed:
feat: generate lang files for item detector covers

mikerooni committed:
feat: fluid detector covers

mikerooni committed:
Merge branch '1.19.2' into detector-covers

mikerooni committed:
style: make method order more consistent in detector covers

mikerooni committed:
chore: regenerate language files to include latest changes

mikerooni committed:
fix: add slight shadows on energy detector cover for better visibility

mikerooni committed:
fix: advanced detector cover UIs now align properly

mikerooni committed:
fix: display correct status when inverting a simple detector cover

mikerooni commented:
In be3a44e, I've added some slight shadows on the energy detector covers (around the needles specifically) for better visibility against the background, as their colors are the only difference between the regular and advanced energy detector.

Please review that as well. I'm not sure whether the shadow is too much or alright like that.

mikerooni committed:
refactor: removed unused field from DetectorCover

Yefancy commented on common/src/main/java/com/gregtechceu/gtceu/utils/TextFormattingUtil.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.stream.Collectors;
+
+public class TextFormattingUtil {
+    public static List<Component> getTranslatableMultilineText(String key) {
+        return getMultilineKeys(key).stream()
+                .map(Component::translatable)
+                .collect(Collectors.toList());
+    }
+
+    private static List<String> getMultilineKeys(String key) {
+        if (!I18n.exists(key + ".0")) {
+            return List.of(key);
+        }
+
+        List<String> keys = new ArrayList<>();
+
+        int index = 0;
+        while (I18n.exists(key + '.' + index)) {
+            keys.add(key + '.' + index);
+            index++;
+        }
+
+        return keys;
+    }
+}
```
why not use the `LangHandler.getMultiLang`

mikerooni replied:
Ahh, I thought `LangHandler` was only for data *generation*. Didn't see that one.

Yefancy commented:
neat lgtm

Yefancy commented on common/src/main/java/com/gregtechceu/gtceu/api/gui/widget/LongInputWidget.java:
```diff
@@ -0,0 +1,150 @@
+    private void increase(ClickData cd) {
+        this.changeValue(cd, 1);
+    }
+
+    private void decrease(ClickData cd) {
+        this.changeValue(cd, -1);
+    }
+
+    private void changeValue(ClickData cd, long multiplier) {
+        if (!cd.isRemote) {
+            long amount = cd.isCtrlClick ? cd.isShiftClick ? 512 : 64 : cd.isShiftClick ? 8 : 1;
+            this.setValue(Mth.clamp(value + (multiplier * amount), min, max));
+        }
+    }
+
+    public LongInputWidget setMin(long min) {
+        this.min = min;
+        updateTextFieldRange();
+
+        return this;
+    }
+
+    public LongInputWidget setMax(long max) {
+        this.max = max;
+        updateTextFieldRange();
+
+        return this;
+    }
+
+    public LongInputWidget setValue(long value) {
+        this.value = value;
+        onChanged.accept(value);
+
+        return this;
+    }
+
+    private void updateTextFieldRange() {
+        this.textField.setNumbersOnly(min, max);
+        this.setValue(Mth.clamp(this.value, min, max));
+    }
```
merge it with your `IntInputWidget`? thry are similar
a `NumberInputWidget` is better.

mikerooni replied:
Hmm, been thinking about a more generic `NumberInputWidget` as well - will try to merge them together.

Yefancy commented on common/src/main/java/com/gregtechceu/gtceu/common/cover/detector/ActivityDetectorCover.java:
```diff
@@ -0,0 +1,41 @@
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public class ActivityDetectorCover extends DetectorCover {
+    @Getter
+    public static final ManagedFieldHolder MANAGED_FIELD_HOLDER = new ManagedFieldHolder(ActivityDetectorCover.class, DetectorCover.MANAGED_FIELD_HOLDER);
```
never be used, there is no managed fields in this class

Yefancy commented on common/src/main/java/com/gregtechceu/gtceu/common/cover/detector/AdvancedEnergyDetectorCover.java:
```diff
@@ -0,0 +1,174 @@
+        return MANAGED_FIELD_HOLDER;
+    }
+
+    @Persisted @DescSynced @Getter @Setter
+    public long minValue, maxValue;
+    @Persisted @DescSynced @Getter @Setter
+    private int outputAmount;
+    @Persisted @DescSynced @Getter
+    private boolean usePercent;
```
do we really need `@DescSynced` here? what are they doing in the remote? 
if its just used for gui, do sync in the gui code is fine.

mikerooni committed:
Revert "refactor: removed unused field from DetectorCover"

This reverts commit 665083296fc7e151ed7291967eb0806f98ddf4d6.

mikerooni committed:
refactor: explicit getter/setter for DetectorCover.isWorkingEnabled

In order to not make the field appear unused, the getter and setter has
been explicitly implemented (instead of Lombok annotations).

mikerooni committed:
refactor: use LangHandler for getting localized multiline strings

mikerooni committed:
refactor: remove managed field holder from ActivityDetectorCover

mikerooni committed:
refactor: remove unnecessary @DescSynced annotations in detector covers

mikerooni committed:
refactor: extract int- and long input widget into common base class

mikerooni committed:
fix: wrong values for energy detector when opening UI in EU mode

Yefancy committed:
fix sync issues + fix drop inner filters
